commands/kubernetes.go

```diff
@@ -395,22 +405,169 @@ func RunKubernetesKubeconfigShow(c *CmdConfig) error {
+}
+
+// loadCachedExecCredential attempts to load the cached exec credential from disk. Never errors
+// Returns nil if there's no credential, if it failed to load it, or if it's expired.
+func loadCachedExecCredential(id string) *clientauthentication.ExecCredential {
+	path := cachedExecCredentialPath(id)
+	f, err := os.Open(path)
+	if err != nil {
+		return nil
+	}
+
+	defer f.Close()
+
+	var execCredential *clientauthentication.ExecCredential
+	if err := json.NewDecoder(f).Decode(&execCredential); err != nil {
+		if Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return nil
+	}
+
+	if execCredential.Status == nil {
+		return nil
+	}
+
+	t := execCredential.Status.ExpirationTimestamp
+	if t.IsZero() || t.Time.Before(time.Now()) {
+		if err := os.Remove(path); err != nil && Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return nil
+	}
+
+	return execCredential
+}
+
+// cacheExecCredential attempts to cache an ExecCredential. Never errors.
+func cacheExecCredential(id string, execCredential *clientauthentication.ExecCredential) {
+	// Don't bother caching if there's no expiration set
+	if execCredential.Status.ExpirationTimestamp.IsZero() {
+		return
+	}
+
+	cachePath := kubeconfigCachePath()
+	if err := os.MkdirAll(cachePath, os.FileMode(0700)); err != nil {
+		if Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return
+	}
+
+	path := cachedExecCredentialPath(id)
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		if Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return
+	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(execCredential); err != nil && Verbose {
+		fmt.Fprintln(os.Stderr, err)
+	}
+}
+
+// RunKubernetesKubeconfigExecCredential displays the exec credential. It is for internal use only.
+func RunKubernetesKubeconfigExecCredential(c *CmdConfig) error {
+	if len(c.Args) != 1 {
+		return doctl.NewMissingArgsErr(c.NS)
+	}
+
+	version, err := c.Doit.GetString(c.NS, doctl.ArgVersion)
+	if err != nil {
+		return err
+	}
+
+	if version != "v1beta1" {
+		return fmt.Errorf("invalid version %q expected 'v1beta1'", version)
+	}
+
+	kube := c.Kubernetes()
+
+	clusterID := c.Args[0]
+	execCredential := loadCachedExecCredential(clusterID)
+	if execCredential != nil {
+		return json.NewEncoder(c.Out).Encode(execCredential)
+	}
+
+	kubeconfig, err := kube.GetKubeConfig(clusterID)
+	if err != nil {
+		return err
+	}
+
+	config, err := clientcmd.Load(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	execCredential, err = execCredentialFromConfig(config)
+	if err != nil {
+		return err
+	}
+	cacheExecCredential(clusterID, execCredential)
+
+	return json.NewEncoder(c.Out).Encode(execCredential)
+}
+
+func execCredentialFromConfig(config *clientcmdapi.Config) (*clientauthentication.ExecCredential, error) {
+	current := config.CurrentContext
+	context, ok := config.Contexts[current]
+	if !ok {
+		return nil, fmt.Errorf("received invalid config from API")
```

Comment:
Error return should be fixed. `fmt.Errorf("context %q doesn't exist", current)`

Reply:
This error is bubbled up to the user, so maybe it should be something user friendly telling the user to file a ticket

Reply:
yeah but how do you know which error it is? Let treat our users more smartly, especially a Kubernetes user would understand what a "missing context" is.

Reply:
No, they would have no idea what to do with it, since the context is missing from the config that is returned by out API, it's not a problem with the configuration of the user. The file that is interpreted here is the kubeconfig that our API provides

So if it can't find the context in this case it means that our API returned a broken kubeconfig

Reply:
gotcha. how do you gonna know during debugging what is wrong? Suppose the user gets the error, is the authinfo missing or there is no context? I think we should at least add something like this: `received invalid config from API (code - 1)` or something alike so we know what it means